cmd/create/create_agent.go

```diff
@@ -0,0 +1,439 @@
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	"k8s.io/kubectl/pkg/cmd/delete"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"time"
+)
+
+const (
+	agentShort   = "Create an agent"
+	agentLong    = "Create an agent"
+	agentExample = `
+	  # Create a new agent
+	  armory create agent`
+
+	defaultNamespaceName = "armory-rna"
+	defaultSecretName    = "rna-client-credentials"
+)
+
+var agentConnectedPollRate = time.Second * 10
+
+type AgentOptions struct {
+	// Name of resource being created
+	Name      string
+	Namespace string
+
+	Context context.Context
+
+	ArmoryClient      *configuration.ConfigClient
+	configuration     *config.Configuration
+	contextNames      []string
+	credentials       *model.Credential
+	kubernetesFactory cmdutil.Factory
+	configAccess      clientcmd.ConfigAccess
+	KubernetesClient  corev1client.CoreV1Interface
+}
+
+// NewAgentOptions creates a new *AgentOptions with sane defaults
+func NewAgentOptions() *AgentOptions {
+	return &AgentOptions{}
+}
+
+func NewCmdCreateAgent(configuration *config.Configuration) *cobra.Command {
+
+	o := NewAgentOptions()
+
+	cmd := &cobra.Command{
+		Use:     "agent",
+		Aliases: []string{},
+		Short:   agentShort,
+		Long:    agentLong,
+		Example: agentExample,
+		RunE: func(cmd *cobra.Command, args []string) error {
+
+			if err := o.Complete(configuration); err != nil {
+				return err
+			}
+
+			if err := o.Validate(); err != nil {
+				return err
+			}
+
+			if err := o.Run(); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+	return cmd
+}
+
+func (o *AgentOptions) Complete(cfg *config.Configuration) error {
+
+	o.configAccess = clientcmd.NewDefaultPathOptions()
+	o.configuration = cfg
+
+	ac := configuration.NewClient(cfg)
+	o.ArmoryClient = ac
+
+	f := o.getKubernetesFactory()
+	o.kubernetesFactory = f
+
+	kc, err := o.getKubernetesClient()
+	if err != nil {
+		return err
+	}
+	o.KubernetesClient = kc
+
+	contextNames, err := o.getContexts()
+	if err != nil {
+		return err
+	}
+	o.contextNames = contextNames
+	o.Context = o.ArmoryClient.ArmoryCloudClient.Context
+	return nil
+}
+
+// Run performs the execution of 'create agent' sub command
+func (o *AgentOptions) Run() error {
+	promptSelectAgent := promptui.Select{
+		Label:  "Please select a context. Your agent will be deployed into the cluster you choose",
+		Items:  o.contextNames,
+		Stdout: &util.BellSkipper{},
+	}
+
+	_, requestedContext, err := promptSelectAgent.Run()
+	if err != nil {
+		return errors.New(fmt.Sprintf("failed to select a context to deploy to; %v\n", err))
+	}
+
+	err = o.useContext(requestedContext)
+	if err != nil {
+		return errors.New(fmt.Sprintf("failed to set context %s; %v\n", requestedContext, err))
+	}
+
+	ctx, cancel := context.WithTimeout(o.Context, time.Minute)
```

Comment:
This explains a behavior which surprised me when I was testing out the CLI. The context created by `context.WithTimeout` will be canceled _one minute from now_, no matter how it's been called. That means that I (as a user) have one minute to finish all of these commands, which is not what you want 😄